assumes/parser.go

```diff
@@ -0,0 +1,257 @@
+	exprTree := struct {
+		AnyOf []ExpressionTree `yaml:"any-of,omitempty"`
+		AllOf []ExpressionTree `yaml:"all-of,omitempty"`
+	}{}
+
+	if err := unmarshalFn(&exprTree); err != nil {
+		if _, isTypeErr := err.(*yaml.TypeError); isTypeErr {
+			return errNotCoercible
+		}
+		return err
+	}
+
+	if len(exprTree.AnyOf)+len(exprTree.AllOf) == 0 {
+		return errors.New(`expected either an "any-of" or "all-of" block`)
+	}
+
+	var boolExpr CompositeExpression
+	var subExprList []ExpressionTree
+
+	if exprTree.AnyOf != nil {
+		subExprList = exprTree.AnyOf
+		boolExpr = CompositeExpression{
+			ExprType: AnyOfExpression,
+		}
+	} else {
+		subExprList = exprTree.AllOf
+		boolExpr = CompositeExpression{
+			ExprType: AllOfExpression,
+		}
+	}
+
+	// Copy the parsed subexpressions to the boolean op
+	boolExpr.SubExpressions = make([]Expression, len(subExprList))
+	for i, subBlock := range subExprList {
+		boolExpr.SubExpressions[i] = subBlock.Expression
+	}
+	tree.Expression = boolExpr
+	return nil
+}
+
+func (tree *ExpressionTree) maybeUnmarshalImplicitAllOfExpr(unmarshalFn func(interface{}) error) error {
+	var subExprList []ExpressionTree
+	if err := unmarshalFn(&subExprList); err != nil {
+		if _, isTypeErr := err.(*yaml.TypeError); isTypeErr {
+			return errNotCoercible
+		}
+
+		return err
+	}
+
+	boolExpr := CompositeExpression{
+		ExprType:       AllOfExpression,
+		SubExpressions: make([]Expression, len(subExprList)),
+	}
+
+	// Copy the parsed subexpressions to the boolean op
+	for i, subBlock := range subExprList {
+		boolExpr.SubExpressions[i] = subBlock.Expression
+	}
+	tree.Expression = boolExpr
+	return nil
+}
+
+func (tree *ExpressionTree) maybeUnmarshalExpr(unmarshalFn func(interface{}) error) error {
+	var exprString string
+	if err := unmarshalFn(&exprString); err != nil {
+		return errNotCoercible
+	}
+
+	expr, err := parseFeatureExprFromString(exprString)
+	if err != nil {
+		return err
+	}
+	tree.Expression = expr
+	return nil
+}
+
+func parseFeatureExprFromString(exprString string) (Expression, error) {
+	exprString = strings.TrimSpace(exprString)
+
+	// Is this a feature name without a version constraint?
+	if featureWithoutVersion.MatchString(exprString) {
+		return FeatureExpression{Name: exprString}, nil
+	}
+
+	matches := featureWithVersion.FindAllStringSubmatch(exprString, 1)
+	if len(matches) == 1 {
+		featName, constraint, versionStr := matches[0][1], matches[0][2], matches[0][3]
+		ver, err := parseVersion(versionStr)
+		if err != nil {
+			return nil, errors.Annotatef(err, "malformed feature expression %q", exprString)
+		}
+
+		return FeatureExpression{
+			Name:       featName,
+			Constraint: VersionConstraint(constraint),
+			Version:    ver,
+			rawVersion: versionStr,
+		}, nil
+	}
+
+	return nil, errors.Errorf("malformed feature expression %q", exprString)
+}
+
+func parseVersion(verStr string) (*version.Number, error) {
+	// Try to parse the version string as-is
+	ver, pErr := version.Parse(verStr)
```

Comment:
I really think we should just fix this. `version.Parse` claims to do something, which it clearly doesn't.

Reply:
While the regex claims it can be used to parse different formats (which is indeed incorrect), the [docblock](https://github.com/juju/version/blob/master/version.go#L162-L164) on `version.Parse` mentions that it assumes `major.minor.patch` format.

As we use that package all over the place I am a bit hesitant to patch that particular method.